Convert detectors to factory pattern, ability to set different model for each detector

dennispg opened the pull request:
Just thought a bit of refactoring could be helpful in adding future detectors.. 

Hope it's helpful!

dennispg committed:
refactor detectors

NickM-27 commented on frigate/object_detection.py:
```diff
@@ -54,19 +52,12 @@ def __init__(
             )
-            self.detect_api = CpuTfl(model_config=model_config, num_threads=num_threads)
+
+        self.detect_api = DetectionApi.create(det_type, det_device=det_device, model_config=model_config, num_threads=num_threads)
```
I'm wondering if there's a better way to do this, if there are detectors in the future that require other args this list could get longer even if most of the args are not relevant to each detector.

dennispg replied:
I was thinking about this as well. I think the best way to do this would be to pass in the `detectors` config as a model, and let the detector itself suss out the args from that. I wasn't sure if I should go that deep into it before knowing whether even this much is beneficial or not though.

NateMeyer replied:
Another thing that was discussed was reorganizing the model configuration to be under the detector, so we can have different models if we have more than one detector type.  That might make it easier to bundle up some of the detector config into a single arg to pass into the create function.

dennispg replied:
Should I try and take at stab at reorganizing the model along with this pull request? Or is it better for a separate pull request?

NickM-27 replied:
Up to you

NickM-27 reviewed (changes requested):
A couple things, and need to take a look at the tests which are failing

dennispg commented on frigate/detectors/tensorrt.py:
```diff
@@ -0,0 +1,19 @@
+import logging
+import numpy as np
+
+from .detection_api import DetectionApi
+from .detector_type import DetectorTypeEnum
+
+
+logger = logging.getLogger(__name__)
+
+
+class TensorRT(DetectionApi):
+    def __init__(self, det_device=None, model_config=None, num_threads=1, **kwargs):
+        raise NotImplementedError("TensorRT engine is not yet functional!")
+
+    def detect_raw(self, tensor_input):
+        raise NotImplementedError("TensorRT engine is not yet functional!")
+
+
+DetectionApi.register_api(DetectorTypeEnum.tensorrt, TensorRT)
```
I started doing this refactor because I was experimenting with a TensorRT detector for Nvidia cards. That didn't really work though, so for the pull request I just wiped it out for an empty implementation. I'll get rid of it altogether, no problem.

dennispg replied:
this is removed now

NateMeyer commented:
Hey, I like this idea.  I think we can improve on the current abstraction implementation with this "factory" design pattern.  It would be great if all the changes for adding a new detector were limited to adding the new implementation python file, and changes to an `__init__.py`.  I'll add a few more comments inline.

NateMeyer commented on frigate/detectors/detector_type.py:
```diff
@@ -0,0 +1,8 @@
+from enum import Enum
+
+
+class DetectorTypeEnum(str, Enum):
```
Can this enum definition be placed right in the __init__ file?

dennispg replied:
this is moved to __init__.py now

NateMeyer commented on frigate/detectors/detection_api.py:
```diff
@@ -8,10 +10,23 @@
+        DetectionApi._api_types[det_type] = det_api
+
+    @staticmethod
+    def create(det_type: DetectorTypeEnum, **kwargs):
```
Should this create function be separate from the DetectionApi class? This is the parent class of the detectors, but it doesn't make sense for each them to have this as a method?

NateMeyer replied:
If this function were defined in the __init__.py file and had access to a static map of the enum to api, I think we could also eliminate the need to have the `register_api` method?

NateMeyer replied:
I slept on it, and I think you may be right.  I think I like this register/create mechanism as-is.

NickM-27 replied:
this can probably be marked as resolved then?

NateMeyer replied:
If I was smart and made these a review instead of comments I could?  I don't get a mark resolved button here though. Lesson learned for next time.

NickM-27 replied:
Weird, I have one so I'll mark it as resolved.

dennispg replied:
oops, I missed the last couple of comments here and made the change already! sorry! I was going back and forth on whether to defend the register/create mechanism as well, so thats why I made the change but I'm going to revert this back.

NickM-27 replied:
@dennispg I wouldn't revert it, I think I like the way this is now TBH. Curious what @NateMeyer thinks

dennispg replied:
Yea, I'll wait for your feedback before I revert. In the meantime, I'm going to take a stab at the model configs

NateMeyer replied:
I think my initial concern was because I was misunderstanding the `@staticmethod` decorator.  I'd be curious to hear your reasons for doing it that way initially.

dennispg replied:
It felt like a concern that is related to the DetectionApi type, but probably the biggest reason is because I'm more of a C# dev than python...

I think I agree with @NickM-27 now though, it feels more "pythonic" now.

NateMeyer commented on frigate/object_detection.py:
```diff
@@ -54,19 +52,12 @@ def __init__(
-                det_device=det_device, model_config=model_config
-            )
-        else:
+        if det_type == DetectorTypeEnum.cpu:
```
Looking at it, this CPU warning should move into the cpu detector? Then we won't need any detector-specific conditions here.

dennispg replied:
the warning is moved into the create_detector function now

dennispg committed:
move create_detector and DetectorTypeEnum

NickM-27 commented:
@dennispg You need to run `python3 -m black frigate` so your code is formatted and python checks pass

dennispg committed:
fixed code formatting

dennispg committed:
add detector model config models

dennispg commented:
I have added some changes for the models that I hope will work. I have several questions/doubts around the unit tests though. 

The biggest one being that I'm not sure what to do about the currently failing tests. There is a reference to a "/test/modelpath" file that doesn't exist anywhere, and I'm actually not sure how that test ever passed even before my changes.

Also there's a few assertions that I think might not be needed, but I'm not sure the original intention behind them. For example, while testing the CPU detector there's an assert that the edgetpu detector wasn't called and vice versa. Probably some work do for those unit tests.

NateMeyer commented:
> I have added some changes for the models that I hope will work. I have several questions/doubts around the unit tests though.
> 
> The biggest one being that I'm not sure what to do about the currently failing tests. There is a reference to a "/test/modelpath" file that doesn't exist anywhere, and I'm actually not sure how that test ever passed even before my changes.
> 
> Also there's a few assertions that I think might not be needed, but I'm not sure the original intention behind them. For example, while testing the CPU detector there's an assert that the edgetpu detector wasn't called and vice versa. Probably some work do for those unit tests.

The purpose of the object detector unit test is to verify that the correct detector is initialized, and that is is passed the correct parameters.  The dummy path `/test/modelpath` is just a test value that is checked, it doesn't need to be real.  The asserts for edgetpu not called are part of testing the correct detector type was created.  The `@patch` decorator is how we setup the a mock interface for the given detector type in this test.

We likely need to tweak the tests with the changes to the passing the model config in the initializer, and the extra layer of the `create_detector` function.  I think we should also add a few tests for the `create_detector` function itself, this could be a new test class.

dennispg commented:
> > I have added some changes for the models that I hope will work. I have several questions/doubts around the unit tests though.
> > The biggest one being that I'm not sure what to do about the currently failing tests. There is a reference to a "/test/modelpath" file that doesn't exist anywhere, and I'm actually not sure how that test ever passed even before my changes.
> > Also there's a few assertions that I think might not be needed, but I'm not sure the original intention behind them. For example, while testing the CPU detector there's an assert that the edgetpu detector wasn't called and vice versa. Probably some work do for those unit tests.
> 
> The purpose of the object detector unit test is to verify that the correct detector is initialized, and that is is passed the correct parameters. The dummy path `/test/modelpath` is just a test value that is checked, it doesn't need to be real. The asserts for edgetpu not called are part of testing the correct detector type was created. The `@patch` decorator is how we setup the a mock interface for the given detector type in this test.
> 
> We likely need to tweak the tests with the changes to the passing the model config in the initializer, and the extra layer of the `create_detector` function. I think we should also add a few tests for the `create_detector` function itself, this could be a new test class.

I agree, I'll get started on some new tests for `create_detector`. I would assume the tests for `create_detector` would obviate the need for those other checks making sure the correct detector is initialized, is that a safe assumption? Because otherwise as more detectors are added, that could grow unweildy. Based on your feedback, I'll clean those up as well. 

But about the dummy path, the tests are still failing for me saying that the file cannot be found. I see that they pass somehow on the dev branch, so I suspect it is something about the way the detectors are now created isn't working with the way they are mocked. I'll check into that, but any pointers would be appreciated there.

For the config models, do those look ok? I left in the global detection model config and fill it into any detectors which don't have their own customized model defined. Kind of like how the cameras can inherit global configuration.

dennispg commented:
I believe the mock patches aren't doing what's expected because the create_detector function is using the dictionary which keeps references to the original unmocked classes. I think fixing that should clear up the failing tests.

NateMeyer commented:
> I agree, I'll get started on some new tests for `create_detector`. I would assume the tests for `create_detector` would obviate the need for those other checks making sure the correct detector is initialized, is that a safe assumption? Because otherwise as more detectors are added, that could grow unwieldy. Based on your feedback, I'll clean those up as well.

Yes, I think you have that right.  

> For the config models, do those look ok? I left in the global detection model config and fill it into any detectors which don't have their own customized model defined. Kind of like how the cameras can inherit global configuration.

I agree it's best if the global model config is still possible.  This will make it easier for people upgrading with an existing configuration.  I'll have to digest the config changes a bit more.

One more thing that might be needed with the model config change, there are a few places we are using the model input format to prepare data for the detector (shape, order, etc).  We may want to have a common method of getting this data when there is a detector-specific model?

> I believe the mock patches aren't doing what's expected because the create_detector function is using the dictionary which keeps references to the original unmocked classes. I think fixing that should clear up the failing tests.

That could cause that issue, if the test is trying to actually create a cpu or edgetpu detector.

dennispg committed:
fix detector unit tests

dennispg commented:
As you pointed out about the model input data for shape, order, and such, I wonder; does it make sense for all of the model's properties to be customizable per detector or a subset, such as the path. I am not 100% familiar with how all that works, so I'm genuinely asking from a place of ignorance.

In [`FrigateApp.start_detectors`](https://github.com/blakeblackshear/frigate/blob/739a2674621a224c6b1f8323c7430dc2d0c16954/frigate/app.py#L184), first it creates some SharedMemory for each camera, and then it creates an `ObjectDetectProcess` for each defined detector. The SharedMemory's size is based on the global model config's height and width right now. Unless we fixed the model height and width globally, wouldn't this mean we would need to create a SharedMemory for each camera/detector combination?

Also, in [`object_processing.py:CameraState` ](https://github.com/blakeblackshear/frigate/blob/739a2674621a224c6b1f8323c7430dc2d0c16954/frigate/object_processing.py#L396)the global model config's colormap is referenced. Could this also probably be fixed globally?

If only some subset of parameters should be customizable per detector, I can create a new DetectorModelConfig class instead of it is right now that contains only that set and fill them in from the global one if empty.

NateMeyer commented:
Yea, I think we shouldn't make any assumptions about the model properties, or that anything would be common between them.

The shm creation brings up a good question, maybe @NickM-27 can weigh in.  If we have multiple detectors available, how does a camera get assigned to a detector? Is it a static mapping? Do we come up with a way to load-balance detectors?

NickM-27 commented:
> The shm creation brings up a good question, maybe @NickM-27 can weigh in. If we have multiple detectors available, how does a camera get assigned to a detector? Is it a static mapping? Do we come up with a way to load-balance detectors?

Currently all the cameras use all the detectors. It is basically a round robin and when a detector finishes an inference it becomes available for another one. I think we definitely would want a way to have all cameras run all detectors as well as a way to pair specific cameras with specific detectors

NateMeyer commented:
> > The shm creation brings up a good question, maybe @NickM-27 can weigh in. If we have multiple detectors available, how does a camera get assigned to a detector? Is it a static mapping? Do we come up with a way to load-balance detectors?
> 
> Currently all the cameras use all the detectors. It is basically a round robin and when a detector finishes an inference it becomes available for another one. I think we definitely would want a way to have all cameras run all detectors as well as a way to pair specific cameras with specific detectors

Would it work just to initialize the shm to the largest model size for now?

NickM-27 commented:
> Would it work just to initialize the shm to the largest model size for now?

That's probably a good way to go about it, AFAIK

dennispg committed:
adjust SharedMemory size to largest detector model shape

dennispg commented:
> > The shm creation brings up a good question, maybe @NickM-27 can weigh in. If we have multiple detectors available, how does a camera get assigned to a detector? Is it a static mapping? Do we come up with a way to load-balance detectors?
> 
> Currently all the cameras use all the detectors. It is basically a round robin and when a detector finishes an inference it becomes available for another one. I think we definitely would want a way to have all cameras run all detectors as well as a way to pair specific cameras with specific detectors


I adjusted the SharedMemory to take the largest model shape as suggested, and fixed a few other places I had missed that needed to pass around the detector configuration. But I'm not sure how best to proceed from here now.

As you pointed out above, currently the detectors are completely decoupled from the frame processing with a queue in between them, and the frame processing has no notion of which detector is going to ultimately do the detection. But it needs that information if the model config is going to be configurable per detector.

Maybe the tensor input creation can be moved in to the LocalObjectDetector.detect phase where it does know it's detector's model config? Seems like that would solve half the problems there, but I think it also could tie up the detector more than necessary and eat away at performance? Also, I only just barely know what I am talking about lol.

And then after detection we need the labelmap from the detector too, but thats also decoupled away.

I feel a bit out of my depth, but maybe with some direction I could keep going. Or, maybe just for now we can just fix the model size and labelmap to be global and deal with making them configurable later?

NickM-27 commented:
> Maybe the tensor input creation can be moved in to the LocalObjectDetector.detect phase where it does know it's detector's model config? Seems like that would solve half the problems there, but I think it also could tie up the detector more than necessary and eat away at performance?

Take a look at the previous discussion https://github.com/blakeblackshear/frigate/pull/3656#issuecomment-1229179521

> I would prefer that each model implementation defines it's own function to prepare the input data in shared memory and then translates the outputs into a common format. Keep in mind that the detection process is extremely sensitive to small increases in processing time. When using a Coral, an extra copy of a frame in memory is enough to increase the total inference times by a significant percentage.

NateMeyer commented:
Have we scope-creeped this change too far?  I think the original detector change can stand alone, with relying on a global model for now.  The detector-specific model configuration could come later to allow multiple mixed-detector support.

NickM-27 commented:
> Have we scope-creeped this change too far?  I think the original detector change can stand alone, with relying on a global model for now.  The detector-specific model configuration could come later to allow multiple mixed-detector support.

I think a bit, I agree it may be better to do in pieces. Might have more clear answers to some of these questions once tesnorrt is implemented

dennispg committed:
fix detector model config defaults

dennispg commented:
I think that is a really good idea.

Since I believe it is still important to be able to configure a custom model path per detector at the minimum, I started down the road of removing the full model config per detector and replacing with just the path. But I got to thinking and went ahead implemented what I hope is something of a compromise instead.

> > Have we scope-creeped this change too far?  I think the original detector change can stand alone, with relying on a global model for now.  The detector-specific model configuration could come later to allow multiple mixed-detector support.
> 
> I think a bit, I agree it may be better to do in pieces. Might have more clear answers to some of these questions once tesnorrt is implemented

I think that is a really good idea!

So, since I believe it is still important to be able to configure a custom model path per detector at the minimum, I started down the road of removing the full model config per detector and replacing it with just the path. But I got to thinking and went ahead implemented what I hope is something of a compromise instead.

I am leaving the full model config per detector in place, merging in any missing config from the global model config, using the detector model config where we can, and leaving the places where we can't as is for now. The compromise is that I am logging an unsupported warning if a user tries to customize any other than the path on the detector model config. (The language of the warning probably needs some word-smithing)

Thoughts?

dennispg committed:
enable auto-discovery of detectors

dennispg committed:
simplify config

dennispg committed:
simplify config changes further

dennispg commented:
I've made some new changes that now allow for a fully pluggable detector mechanism. To add a detector, you need only add a module under `detectors/plugins` with a class that inherits from `DetectionApi` and exposes a string attribute `type_key`. No other changes anywhere else required, everything should just pick up from there.

I'll leave the re-architecting to allow for more fully configurable detector specfic model configurations for a later time.

If this is acceptable, should I go ahead and add to the docs about being able to configure the model path on the detector?

NickM-27 commented:
I'd go ahead and do the docs 👍

dennispg committed:
update detectors docs; detect detector configs dynamic

dennispg commented:
Finished the docs!

NickM-27 commented on docs/docs/configuration/detectors.md:
```diff
@@ -3,11 +3,37 @@ id: detectors
+
+When using CPU detectors, you can add one CPU detector per camera. Adding more detectors than the number of cameras should not improve performance.
+
+## Edge-TPU Detector
+The EdgeTPU detector type runs a TensorFlow Lite model utilizing the Google Coral delegate for hardware acceleration. To configure an EdgeTPU detector, set the `"type"` attribute to `"edgetpu"`.
```
```suggestion
## Edge-TPU Detector

The EdgeTPU detector type runs a TensorFlow Lite model utilizing the Google Coral delegate for hardware acceleration. To configure an EdgeTPU detector, set the `"type"` attribute to `"edgetpu"`.
```

NickM-27 commented on docs/docs/configuration/detectors.md:
````diff
@@ -64,38 +92,32 @@ detectors:
-```
-
-When using CPU detectors, you can add a CPU detector per camera. Adding more detectors than the number of cameras should not improve performance.
+## OpenVINO Detector
+The OpenVINO detector type runs an OpenVINO IR model on Intel CPU, GPU and VPU hardware. To configure an OpenVINO detector, set the `"type"` attribute to `"openvino"`.
````
```suggestion
## OpenVINO Detector

The OpenVINO detector type runs an OpenVINO IR model on Intel CPU, GPU and VPU hardware. To configure an OpenVINO detector, set the `"type"` attribute to `"openvino"`.
```

NickM-27 commented on docs/docs/configuration/index.md:
```diff
@@ -74,15 +74,13 @@ mqtt:
-    # Optional: num_threads value passed to the tflite.Interpreter (default: shown below)
-    # This value is only used for CPU types
-    num_threads: 3
+    # Frigate provided types include 'cpu', 'edgetpu', and 'openvino'
```
```suggestion
    # Frigate provided types include 'cpu', 'edgetpu', and 'openvino' (default: shown below)
```

NickM-27 commented on frigate/app.py:
```diff
@@ -186,10 +186,16 @@ def start_detectors(self) -> None:
             self.detection_out_events[name] = mp.Event()
 
             try:
+                size = max(
+                    [
+                        det.model.height * det.model.width * 3
+                        for (name, det) in self.config.detectors.items()
+                    ]
+                )
```
```suggestion
                largest_frame = max(
                    [
                        det.model.height * det.model.width * 3
                        for (name, det) in self.config.detectors.items()
                    ]
                )
```

NickM-27 commented on frigate/app.py:
```diff
@@ -186,10 +186,16 @@ def start_detectors(self) -> None:
                     name=name,
                     create=True,
-                    size=self.config.model.height * self.config.model.width * 3,
+                    size=size,
```
```suggestion
                    size=largest_frame,
```

NickM-27 commented on frigate/config.py:
```diff
@@ -890,7 +834,7 @@ class FrigateConfig(FrigateBaseModel):
     )
     detectors: Dict[str, DetectorConfig] = Field(
-        default={name: DetectorConfig(**d) for name, d in DEFAULT_DETECTORS.items()},
+        default=parse_obj_as(Dict[str, DetectorConfig], DEFAULT_DETECTORS),
```
I think this can just be DEFAULT_DETECTORS, no need for parse_obj_as in the default

dennispg replied:
all the above suggestions are implemented. I also added some docs on how to provide a custom implementation that I'm pretty sure will need some word-smithing.

dennispg committed:
add suggested changes

NickM-27 commented on docs/docs/configuration/detectors.md:
````diff
@@ -145,3 +147,38 @@ device_cgroup_rules:
+
+    def detect_raw(self, tensor_input):
+      return <list of detections>
+```
````
This doesn't belong here. These are the user docs, this is something that may belong in the contribution section

dennispg replied:
My thought process on it is that it could enable users to provide their own implementations by just bind mounting a .py file in, if they so desired. It's an advanced scenario sure, but not implausible?

NickM-27 replied:
This isn't something that I think we would want to support officially though. If someone wants to add a new detector in a PR happy to help but, we don't want a user expecting support in doing so. 

And a user adding this advanced of a feature is a "developer" anyway

NickM-27 replied:
Like I said I think it's fine I'm the contributions section of the docs, just not the user part 👍

dennispg committed:
remove custom detector doc

dennispg committed:
fix grammar, adjust device defaults

herostrat commented:
Hey @dennispg, I really like this refactoring.

Just brainstorming here (which is probably a bit/a lot too outside of the box). For external/manuel trigger (#1063 or #3184) I think your plugins would be a good fit.
E.g. a external_trigger (mqtt, web api) or audio_trigger plugin(detection in cams audio stream)?
Is this too far fetched? On first sight I think this would be doable and kind of reasonable from an architecture point of view?

dennispg commented:
> Hey @dennispg, I really like this refactoring.
> 
> Just brainstorming here (which is probably a bit/a lot too outside of the box). For external/manuel trigger (#1063 or #3184) I think your plugins would be a good fit. E.g. a external_trigger (mqtt, web api) or audio_trigger plugin(detection in cams audio stream)? Is this too far fetched? On first sight I think this would be doable and kind of reasonable from an architecture point of view?

To be fully honest, I only very recently started getting aquainted with this codebase. Have been using it for a while at home, but with detection turned off since I couldn't find a Coral. I had been hoping that a way to use my GPU would eventually be implemented, but finally decided to poke around and tinker myself to see if there's anything I could contribute. That's what led me to doing this. 

All that just to say that I'm not sure I'm qualified to give a proper answer with my limited exposure, but still, I think that what I've done here is too specific to just the detectors and I don't think it would translate well to what you're suggesting.

dennispg commented:
@NickM-27 @NateMeyer I am toying with a mechanism that builds on this pull request which enables detection requests to be sent off to a remote server for processing. 

I have been playing around with trying to get TensorRT working, but going in circles with trying to get a container image built with the proper dependencies. I was thinking to set it up as a seperate tag or something so as not to clutter up the main image with too much junk you may not need...

WIth this in place, you can build an image with TensorRT from Nvidia's base images that only runs a server listening for requests for detection. But more than that, it enables being able to use many machines together to distribute detection as well.

I've kind of got the client/server mechanism working now, and would love to run it by you guys for some feedback. Since it builds on this pull request though, should I wait till this is merged and start a new pull request? Or maybe just add the commits here?

NateMeyer commented:
@dennispg I have also been playing with updating the tensorRT patches the last couple weeks. I think I have the libraries in place, but agree that it really bloats the amd64 image if you aren't using it.  I was considering how to make it a separate component along the same lines you are thinking, but with using the shm to move data between containers.  Although I am weary that such a setup would be too difficult for someone who isn't an experienced docker user.

That said, my first goal is to get something working on the amd64 image, then we can optimize it from there.  I'll push what I have to GitHub tomorrow if you want to take a look. I'm still going through and fixing some cuda memory buffers and it doesn't yet run without crashing.

I would keep this PR focused to the detector changes you have here.

NickM-27 commented:
> I'll push what I have to GitHub tomorrow if you want to take a look

Looking forward to it

> I would keep this PR focused to the detector changes you have here.

100% agree I think this PR should stay focused on what it is now

dennispg commented:
@NateMeyer I actually did also built in an option to share the data either via shared memory or via streaming the bytes over a socket. The socket is much easier to configure and can be used even across machines, but slower. The shm requires to be on the same machine, but like you said, maybe not so straightforward to set up for a regular user.

Anyways, I agree with you both too keep this PR focused. I will push what I have to a branch in my fork soon and then maybe take a look from there. I'd love to take a look at your TensorRT work also once you've pushed that.

blakeblackshear commented:
I had considered TensorRT in a separate container with shared shm when I looked into it originally a long time ago. It all got so complicated that I decided it wasn't worth the extra complexity. Not saying to stop going down that path, but we will have to think carefully about any additional support burdens it could bring. Also, @NickM-27 and I have been discussing adding zeromq in a future version to replace all the interprocess communication and allow certain processes to be written in more performant languages where beneficial. That could make sense for things like this.
